src/auto4_lua.cpp

```diff
@@ -623,27 +623,33 @@ namespace {
 	{
 		bool failed = false;
 		BackgroundScriptRunner bsr(parent, title);
-		bsr.Run([&](ProgressSink *ps) {
-			LuaProgressSink lps(L, ps, can_open_config);
-
-			// Insert our error handler under the function to call
-			lua_pushcclosure(L, add_stack_trace, 0);
-			lua_insert(L, -nargs - 2);
-
-			if (lua_pcall(L, nargs, nresults, -nargs - 2)) {
-				if (!lua_isnil(L, -1)) {
-					// if the call failed, log the error here
-					ps->Log("\n\nLua reported a runtime error:\n");
-					ps->Log(get_string_or_default(L, -1));
+		try {
+			bsr.Run([&](ProgressSink *ps) {
+				LuaProgressSink lps(L, ps, can_open_config);
+
+				// Insert our error handler under the function to call
+				lua_pushcclosure(L, add_stack_trace, 0);
+				lua_insert(L, -nargs - 2);
+
+				if (lua_pcall(L, nargs, nresults, -nargs - 2)) {
+					if (!lua_isnil(L, -1)) {
+						// if the call failed, log the error here
+						ps->Log("\n\nLua reported a runtime error:\n");
+						ps->Log(get_string_or_default(L, -1));
+					}
+					lua_pop(L, 2);
+					failed = true;
 				}
-				lua_pop(L, 2);
-				failed = true;
-			}
-			else
-				lua_remove(L, -nresults - 1);
+				else
+					lua_remove(L, -nresults - 1);
 
-			lua_gc(L, LUA_GCCOLLECT, 0);
-		});
+				lua_gc(L, LUA_GCCOLLECT, 0);
+			});
+		} catch (agi::UserCancelException const&) {
+			if (!failed)
+				lua_pop(L, 2);
```

Comment:
Since this exists outside of the bsr, can something else modify the lua stack after the exception is thrown? Basically, is this actually safe to just pop?

Reply:
As far as I can see, this is safe. The `UserCancelException` is only ever thrown after the entire function passed to `bsr.Run` runs through, so no lua code should be running at that point and we know how the lua stack looks.

That said, when revisiting this I did notice that this was incorrect in the case of export filters (which have a different number of returned values), and the actual number of values to be popped here is `nresults`. I updated this now and added a comment.